### ISSUE TITLE
aiida-qe bands workflow: process shown as "waiting"

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida-tutorials#123]
While running the aiida-qe bands workflow, the PwCalculation is shown as "Waiting for transport task: update".

```
$ verdi process list
  PK  Created    Process label     Process State    Process status
----  ---------  ----------------  ---------------  ----------------------------------
 189  4m ago     PwBandsWorkChain  ⏵ Waiting        Waiting for child processes: 191
 191  4m ago     PwRelaxWorkChain  ⏵ Waiting        Waiting for child processes: 194
 194  4m ago     PwBaseWorkChain   ⏵ Waiting        Waiting for child processes: 209
 209  1m ago     PwCalculation     ⏵ Waiting        Waiting for transport task: update

Total results: 4
```

This is running on a `local` pw.x executable (and `top` shows that `pw.x` is correctly running).

@sphuber @borellim 
This might be an aiida-core issue?

**Comment by sphuber** [comment]
Do they all eventually finish and end up in `Finished`? If so, I am not sure what the problem is that you are alluding to. Is it the fact that the `PwCalculation` is in `Waiting` while it is waiting for an update?

**Comment by ltalirz** [comment]
> Is it the fact that the PwCalculation is in Waiting while it is waiting for an update?

Well yes - I might be a bit behind but what happened to `running` calculations?
As a new user, I would expect to see at least one calculation running at any point in time?

The message "Waiting for transport task: update" does not tell me that this calculation is running.

**Comment by sphuber** [comment]
`Running` is a process state and so applies to the `Process`. Just like the work chains are `Waiting` because the processes are not actually being "run" currently by the engine. They are all waiting for something else. In the case of the `CalcJob` it is waiting for an update of the status of the job on the computer.

**Comment by ltalirz** [comment]
Yes, I understand the logic and it is internally consistent - still, what matters is the message/information passed to the user. 

In this logic, a `CalcJob` will be shown as "running" only during parsing/copying steps, which is exactly the times during which the actual calculation job is *not* running.
I think this will be very counterintuitive to most new users and it would be nice if we could come up with a message that can be understood by people without the need to be familiar with implementation details (that we run a local process per calculation).

I'll think a bit about this; mentioning @giovannipizzi for comments/ideas as well

**Comment by giovannipizzi** [comment]
I tend to agree with @ltalirz's comment.

In earlier versions `verdi calculation list` had also the scheduler state, that is much more intuitive to users for CalcJobs. I think we should find some easy way to bring this back for CalcJobs, or some similar way to convey what is really going on/what is "running" in a more general and less AiiDA-centric point of view.

**Comment by sphuber** [comment]
I'm in principle fine by trying to make it clearer, but I want to avoid that this goes too much at the cost of consistency and code clarity. I don't think that ultimately it is a good thing if we are going to add exceptions and different behaviors for calculation jobs, just because we are traditionally most used to that type of calculation

**Comment by giovannipizzi** [comment]
One solution could be to have a `verdi calcjob list` that is similar but specific only to calcjobs, and with info related to it (in the end, it's what the old `verdi calculation list` was).

**Comment by sphuber** [comment]
If all that is going to add is a column with the scheduler state (not sure what else there is to show), I am not sure if that is going to help new users. Then you have to use two commands instead of one and explain why there is a difference.

**Comment by giovannipizzi** [comment]
Another option: improve the message from
`Waiting for transport task: update`
to something more like `Waiting for scheduled job to complete (current state: QUEUED/RUNNING/UNKNOWN/...) [transport task: update]`
(I know it's long but just to give the idea)

**Comment by sphuber** [comment]
How about: `Waiting for transport task: update (current job state 'RUNNING')`. By simply appending the scheduler state at least it remains consistent with other messages like `Waiting for transport task: submit`

**Comment by giovannipizzi** [comment]
ok for me!

**Comment by ltalirz** [comment]
> `Waiting for transport task: update (current job state 'RUNNING')`

I think this would already be a significant improvement. Since this is quite long (see example above), perhaps rather `Waiting for transport task: update (job 'RUNNING')`  (which might still be too long...).

```
$ verdi process list
  PK  Created    Process label     Process State    Process status
----  ---------  ----------------  ---------------  ----------------------------------
 189  4m ago     PwBandsWorkChain  ⏵ Waiting        Waiting for child processes: 191
 191  4m ago     PwRelaxWorkChain  ⏵ Waiting        Waiting for child processes: 194
 194  4m ago     PwBaseWorkChain   ⏵ Waiting        Waiting for child processes: 209
 209  1m ago     PwCalculation     ⏵ Waiting        Waiting for transport task: update (job RUNNING)

Total results: 4
```

Even with this improvement, however, I still feel it is quite difficult to see at a glance which jobs are actually running.
You can't tell by the state (workchains & jobs are both "Waiting"), and in the status message the relevant information appears at the end, rather than at the beginning.

Perhaps it would make sense to customize the messages for the different "transport tasks" - i.e. rather than prefixing all of them by `Waiting for transport task`, have full specific messages.

Such as `Waiting for scheduler update (job RUNNING)`

**Comment by sphuber** [comment]
Ok, so let's try to complete this table:

Task | Process status
------ | ---------------------
`UPLOAD` | `Waiting for transport task: upload`
`SUBMIT` | `Waiting for transport task: submit`
`UPDATE` |  `Awaiting scheduler update (SCHEDULER_STATE)`
`RETRIEVE` |  `Waiting for transport task: retrieve`

Suggestions for the other steps?

**Comment by sphuber** [comment]
> And I think the point remains that the Process State column does not really add much useful information when looking at "running" workchains.

It does however, when you have paused processes or are looking also at terminated processes. Also if you have long running process functions, they will be shown as `Running` and they don't have a process status, so there the column is also important and informative

**Comment by ltalirz** [comment]
> Ok, so let's try to complete this table:

One question: What would be the reasons for jobs to wait for UPLOAD/SUBMIT/RETRIEVE?
Is it just the cooldown interval (i.e. are they waiting for the next possible connection) or can there be other reasons?
I.e. will jobs likely be in these states for longer periods of time or not?

In general, I think the messages for UPLOAD/SUBMIT/RETRIEVE are less of a problem since there the idea of them "waiting to start doing X" coincides with the intuition of the user.

> It does however, when you have paused processes or are looking also at terminated processes. Also if you have long running process functions, they will be shown as Running and they don't have a process status, so there the column is also important and informative.

That's certainly correct, and I'm not saying it isn't useful in general. 
I just think that in the process list above, a user would like to be able to tell in one glance "where the action is".

**Comment by sphuber** [comment]
Correct, for the upload/submit/retrieve steps, currently, the only reason for waiting is for them to wait on the transport to become available. However, this will typically be at least 60 seconds for remote clusters. Any shorter intervals risk to bombard the cluster too much

**Comment by ltalirz** [comment]
> Correct, for the upload/submit/retrieve steps, currently, the only reason for waiting is for them to wait on the transport to become available.

Ok. So, I guess this means there is a distinction between UPDATE and the other tasks in that UPDATE is repetitive while the others aren't.

I've been thinking a bit more about `Awaiting scheduler update (RUNNING)` and I think it can be misread to mean "scheduler update is running", so I think there is no way around putting "job" in there somewhere (which makes things too long).

Would it be OK to remove the "Waiting" from this process status? I.e. simply to say 
`Monitoring scheduler (job RUNNING)`.

```
$ verdi process list
  PK  Created    Process label     Process State    Process status
----  ---------  ----------------  ---------------  ----------------------------------
 189  4m ago     PwBandsWorkChain  ⏵ Waiting        Waiting for child processes: 191
 191  4m ago     PwRelaxWorkChain  ⏵ Waiting        Waiting for child processes: 194
 194  4m ago     PwBaseWorkChain   ⏵ Waiting        Waiting for child processes: 209
 209  1m ago     PwCalculation     ⏵ Waiting        Monitoring scheduler (job RUNNING)

Total results: 4
```

I think this would be fine, since it's anyhow just a message for the user - from an API perspective you can still use the Process state which will be `Waiting`.
For me, the messages for UPLOAD/SUBMIT/RETRIEVE are fine.

The only question remaining would be what to display what process status to display when the PwCalculation process is actually RUNNING the update (perhaps something like `updating job status`). Anyhow I guess this would only be shown for very short periods of time.

**Comment by sphuber** [comment]
> The only question remaining would be what to display what process status to display when the PwCalculation process is actually RUNNING the update (perhaps something like `updating job status`). Anyhow I guess this would only be shown for very short periods of time.

You are right, it will be shown relatively shortly as this is the point that the engine is actually running the process. Typically those actions will be fast but it depends. The upload, retrieval and parsing can take longer. In my experience, however, I have never had trouble here, nor have we heard from other users. Let's wait to see if we get comments about this before changing anything.

**Comment by giovannipizzi** [comment]
I think the latest suggestion
```
209  1m ago     PwCalculation     ⏵ Waiting        Monitoring scheduler (job RUNNING)
```
improves things a lot.
I would suggest to try to push this for 1.0 since I think this should be easy to implement. We can explicitly ask some user (ideally not too exposed to AiiDA) what they think it's happening if they see the current version:
```
$ verdi process list
  PK  Created    Process label     Process State    Process status
----  ---------  ----------------  ---------------  ----------------------------------
 189  4m ago     PwBandsWorkChain  ⏵ Waiting        Waiting for child processes: 191
 191  4m ago     PwRelaxWorkChain  ⏵ Waiting        Waiting for child processes: 194
 194  4m ago     PwBaseWorkChain   ⏵ Waiting        Waiting for child processes: 209
 209  1m ago     PwCalculation     ⏵ Waiting        Waiting for transport task: update

Total results: 4
``` 
and then without explaining what the right answer is, asking what is going on in the new proposal. So depending on the answers we will understand if the change is useful (personally, I vote to change it)

**Comment by giovannipizzi** [comment]
(Also, we might want to move this issue to aiida-core (with a different title)?)
